Sources/_MatchingEngine/Regex/Parse/Mocking.swift

```diff
@@ -1,107 +1,142 @@
 
-/// Mock-up of the Swift lexer, so we can model our interfaces
-struct SwiftLexer {
-  enum Tok {
-    case opaque(AST)
-
-    var ast: AST {
-      switch self {
-      case .opaque(let ast): return ast
-      }
-    }
-  }
+// TODO: mock up multi-line soon
 
-  static func lex(_ src: String) -> Tok {
-    // Assume there's just a single literal here, no escaping
-    if let start = src.firstIndex(of: "'") {
-      return lexSingleQuotedLiteral(
-        src[src.index(after: start)..<src.lastIndex(of: "'")!])
+public enum Delimiter: Hashable, CaseIterable {
+  case regular
+  case modern
+
+  var openingAndClosing: (opening: String, closing: String) {
+    switch self {
+    case .regular: return ("'/", "/'")
+    case .modern: return ("'|", "|'")
     }
-    fatalError()
   }
-
-  static func lexSingleQuotedLiteral(_ sub: Substring) -> Tok {
-    let ast: AST
-    if sub.first == "/" && sub.last == "/" {
-      ast = try! parse(
-        sub.dropFirst().dropLast(), .traditional)
-    } else {
-      ast = try! parse(sub, .modern)
+  public var opening: String { openingAndClosing.opening }
+  public var closing: String { openingAndClosing.closing }
+
+  /// The default set of syntax options that the delimiter indicates.
+  public var defaultSyntaxOptions: SyntaxOptions {
+    switch self {
+    case .regular: return .traditional
+    case .modern: return .modern
     }
+  }
+}
 
-    // TODO: ... call regex parser
-    return .opaque(ast)
+public struct LexError: Error, CustomStringConvertible {
+  enum Kind: Hashable {
+    case endOfString
+    case invalidUTF8 // TODO: better range reporting
+    case unknownDelimiter
   }
 
-}
+  var kind: Kind
 
-// TODO: mock up multi-line soon
+  /// The pointer at which to resume lexing.
+  public var resumePtr: UnsafeRawPointer
```

Comment:
So we have lexing errors and parsing errors. I thought a lexing error would just send the Swift compiler down the diagnosing single quotes error path, or even better, a path that checks the experimental flag and tells you what delimiters to use.

What is the resumption design?

Reply:
> I thought a lexing error would just send the Swift compiler down the diagnosing single quotes error path

For errors like invalid UTF-8 and end-of-string, wouldn't we want to diagnose that instead?

> What is the resumption design?

After emitting the error, the C++ lexer resumes lexing at the given pointer, which lets us have better error handling for things like an unterminated literal, as we'd resume lexing at the newline.

Reply:
Ah makes sense.

> For errors like invalid UTF-8 and end-of-string, wouldn't we want to diagnose that instead?

We could emit it, or we could fall back to the lexer to emit it. I thought the latter might be a simpler integration story for now, but either way.